validator/core/src/output/data_type.rs

```diff
@@ -0,0 +1,356 @@
+
+    /// Type parameters for non-simple types.
+    pub parameters: Vec<Parameter>,
+}
+
+impl std::fmt::Display for DataType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.class)?;
+        if self.nullable {
+            write!(f, "?")?;
+        }
+        if let Some(variation) = &self.variation {
+            write!(f, "[{}]", variation)?;
+        }
+        if self.parameters.is_empty() {
+            write!(f, "<")?;
+            let mut first = true;
+            for parameter in self.parameters.iter() {
+                if first {
+                    first = false;
+                } else {
+                    write!(f, ", ")?;
+                }
+                write!(f, "{}", parameter)?;
+            }
+            write!(f, ">")?;
+        }
+        Ok(())
+    }
+}
+
+/// Trait for things that can resolve user-defined types and type variations.
+pub trait TypeResolver {
+    /// Resolves a user-defined type from its name.
+    fn resolve_type<S: AsRef<str>>(&self, s: S) -> diagnostic::Result<Arc<extension::DataType>>;
+
+    /// Resolves a type variation from its name and base type.
+    fn resolve_type_variation<S: AsRef<str>>(
+        &self,
+        s: S,
+        base_type: Class,
+    ) -> diagnostic::Result<Arc<extension::TypeVariation>>;
+}
+
+/// Trait for checking the type parameters for a base type.
+trait ParameterChecker {
+    /// Checks whether the given parameter set is valid for this base type.
+    fn check_parameters(&self, params: &[Parameter]) -> diagnostic::Result<()>;
+}
+
+impl DataType {
+    /// Parse a string as a type.
+    pub fn parse<S: AsRef<str>, R: TypeResolver>(
+        _s: S,
+        _type_resolver: R,
+    ) -> diagnostic::Result<DataType> {
+        todo!(
+            "use nom or some other parser to implement this;
+            also run ParameterChecker. then make round-trip tests.
+            this should probably also not return Result; the
+            prototype should allow returning some best-effort/error
+            type for recovery in addition to any number of diagnostics"
+        )
+    }
+}
+
+/// Type class.
+#[derive(Clone, Debug, PartialEq)]
+pub enum Class {
+    /// Well-known simple type.
+    Simple(Simple),
+
+    /// Well-known compound type.
+    Compound(Compound),
+
+    /// User-defined type.
+    UserDefined(Arc<extension::Reference<extension::DataType>>),
+
+    /// Unresolved type. Used for error recovery.
+    Unresolved(String),
+}
+
+impl std::fmt::Display for Class {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Class::Simple(simple) => write!(f, "{}", simple),
+            Class::Compound(compound) => write!(f, "{}", compound),
+            Class::UserDefined(user_defined) => write!(f, "{}", user_defined),
+            Class::Unresolved(name) => write!(f, "{}!", name),
+        }
```

Comment:
```suggestion
       write!(f, "{}", match self {
            Class::Simple(simple) => simple,
            Class::Compound(compound) => compound,
            Class::UserDefined(user_defined) => user_defined,
            Class::Unresolved(name) => name,
        })
```

Reply:
Doesn't work because simple, compound, user_defined, and name are all different types.